spec/controllers/automated_tests_controller_spec.rb

```diff
@@ -133,33 +133,44 @@
       before { post_as user, :upload_files, params: params }
       after { FileUtils.rm_r assignment.autotest_files_dir }
       context 'uploading a zip file' do
-        let(:zip_file) { fixture_file_upload(File.join('/files', 'test_zip.zip'), 'application/zip') }
+        let(:zip_file1) { fixture_file_upload(File.join('/files', 'test_zip.zip'), 'application/zip') }
+        let(:zip_file2) { fixture_file_upload(File.join('/files', 'zip_file.zip'), 'application/zip') }
         let(:unzip) { 'true' }
-        let(:params) { { assignment_id: assignment.id, unzip: unzip, new_files: [zip_file], path: '' } }
+        let(:params) { { assignment_id: assignment.id, unzip: unzip, new_files: [zip_file1, zip_file2], path: '' } }
         let(:tree) { assignment.autotest_files }
         context 'when unzip if false' do
           let(:unzip) { 'false' }
           it 'should just upload the zip file as is' do
             expect(tree).to include('test_zip.zip')
           end
           it 'should not upload any other files' do
-            expect(tree.length).to eq 1
+            expect(tree.length).to eq 2
           end
         end
-        it 'should not upload the zip file' do
-          expect(tree).not_to include('test_zip.zip')
-        end
-        it 'should upload the outer dir' do
-          expect(tree).to include('test_zip')
-        end
-        it 'should upload the inner dir' do
-          expect(tree).to include('test_zip/zip_subdir')
-        end
-        it 'should upload a file in the outer dir' do
-          expect(tree).to include('test_zip/Shapes.java')
+        context 'When the file is a zip file containing sub directory and files inside sub directory' do
+          it 'should not upload the zip file' do
+            expect(tree).not_to include('test_zip.zip')
+          end
+          it 'should upload the outer dir' do
+            expect(tree).to include('test_zip')
+          end
+          it 'should upload the inner dir' do
+            expect(tree).to include('test_zip/zip_subdir')
+          end
+          it 'should upload a file in the outer dir' do
+            expect(tree).to include('test_zip/Shapes.java')
+          end
+          it 'should upload a file in the inner dir' do
+            expect(tree).to include('test_zip/zip_subdir/TestShapes.java')
+          end
         end
-        it 'should upload a file in the inner dir' do
-          expect(tree).to include('test_zip/zip_subdir/TestShapes.java')
+        context 'When the file is a zip file without any sub directory' do
```

Comment:
The issue does _not_ occur when there are _no_ subdirectories. The issue occurs when the zip file contains an entry for a file in a subdirectory but no entry for the subdirectory itself. This does not test that case.